### PR TITLE
fix(studio+mcp): the demo-hardening batch behind the Studio clips

### PR DESCRIPTION
**Pull request by vboussot**
Everything the filming exercise surfaced, 47 commits.

Product fixes:
- wait_for_job reports a timeout as a payload (timed_out, waited_s, next_actions) instead of raising: long trainings no longer fill the chat with error cards.
- The chat transcript and history live under .konfai_studio/sessions/<name>/, outside the konfai-mcp workspace: initialize_session(overwrite) can no longer destroy the conversation that asked for it. Legacy layouts are adopted.
- Session pinning, criterion device placement, Transform.h5 reading, leaderboard skeleton, tab hover ground, and the smaller frictions the exercise surfaced.
- run_transform accepts a gpu list like every other runner (the agent itself reported the gap on camera).
- Bounded CPU thread share per rank and the transform device plumbing, duplicated here from #103 and #104 so the branch stands alone; merging those first drops the duplicates.

Validated by driving Studio black-box through five filmed end-to-end sessions: train/predict/evaluate, live lr tuning and stop, a published app with Dice evaluation, a custom loss authored and smoke-tested by the agent, and a 75-case TRANSFORM under a 512 MiB budget.

## Summary by CodeRabbit

- **New Features**
  - Fine-tuning supports configurable batch sizes through CLI, API, and remote workflows.
  - Transform jobs support GPU selection, compact summaries, and full plans in logs.
  - ITK transform files provide structured summaries and displacement statistics.
  - Studio adds LLM connection management, 3D Slicer integration, folder navigation, live run status, and improved experiment controls.

- **Bug Fixes**
  - Improved session protection, timeout reporting, dataset path handling, stream resilience, save conflict detection, and TensorBoard startup failures.

- **Documentation**
  - Updated transform outputs, planning, fine-tuning options, Studio configuration, and tool behavior documentation.

### CODE REVIEW

File: docs/source/usage/making-data.md [hunk 1]
```diff
@@ -150,9 +150,8 @@ flowchart TB
 `SKIP` means the run resumes, `REDUCE` folds the cohort, `WHOLE-VOLUME` names the
 stage that refused to stream, `STREAM` reads region by region.
 `LOAD` is a cost decision, not a failure: past a threshold, reading the case once
-beats reading regions of it many times. A run writes its plan to
-`Transforms/<name>/plan.txt` next to an `outputs.json` saying where each
-deliverable landed.
+beats reading regions of it many times. A run's log opens with its plan, next to an
+`outputs.json` saying where each deliverable landed.
```
[the remark this answers is not in the document]
Comment (vboussot) [reply docs/source/usage/making-data.md]: Fixed in 7a3b2a1: both pages describe outputs.json as declaring each configured Write destination.

File: konfai-apps/konfai_apps/app_repository.py [hunk 2]
```diff
@@ -203,6 +203,16 @@ class TerminologyEntry:
...
+# Training knobs a caller names directly, and the config key each one writes. Sugar over ``--set``:
+# one path applies them, so a knob costs a row here rather than its own YAML-writing branch.
+_TRAINING_KNOBS = {"batch_size": "Trainer.Dataset.batch_size"}
+
+# A bare ``--set`` name resolves inside the model block only; every other key needs its full path.
+_DOTTED_PATH_HINT = (
+    "A bare name is a model parameter; a config key needs its full dotted path from the config root,"
+    " e.g. 'Trainer.Dataset.batch_size=4'."
+)
+
```
[the remark this answers is not in the document]
Comment (vboussot) [reply konfai-apps/konfai_apps/app_repository.py:217]: Fixed in 7b2f851: the hint names both roots (Predictor... for an inference config, Trainer... for a fine-tune one).

File: konfai-apps/konfai_apps/app_repository.py [hunk 3]
```diff
@@ -1115,16 +1125,18 @@ def install_fine_tune(
...
         Shared assets (config, code, ``app.json``, ``requirements.txt``) are installed into
         ``path``; ``app.json`` is rewritten with the new ``display_name`` and a ``models`` list
         limited to the selected checkpoints, and the training config's ``epochs``/``it_validation``
-        are updated. ``overrides`` are ``--set NAME=VALUE`` model/config tweaks baked into the training
-        config before training (same syntax as :meth:`_apply_config_overrides`). Only the selected ``.pt``
-        checkpoints are downloaded. Returns ``(basename, source_path)`` pairs for the checkpoints to fine-tune.
+        (and ``batch_size``, when given) are updated. ``overrides`` are ``--set NAME=VALUE`` model/config
+        tweaks baked into the training config before training (same syntax as
+        :meth:`_apply_config_overrides`). Only the selected ``.pt`` checkpoints are downloaded. Returns
+        ``(basename, source_path)`` pairs for the checkpoints to fine-tune.
```
[the remark this answers is not in the document]
Comment (vboussot) [reply konfai-apps/konfai_apps/app_repository.py:1141]: Fixed in 7b2f851: install_fine_tune raises AppRepositoryError for batch_size <= 0 before installing anything; zero and negative are covered by a parametrized test.

File: konfai-mcp/konfai_mcp/file_io.py [hunk 4]
```diff
@@ -59,6 +59,71 @@ def read_text_range(path: Path, max_chars: int = 20000, offset: int = 0) -> dict
...
+                        import numpy as np
+
+                        vectors = np.asarray(params, dtype=float).reshape(-1, 3)
+                        magnitudes = np.linalg.norm(vectors, axis=1)
+                        entry["displacement_summary"] = {
+                            "vectors": int(vectors.shape[0]),
+                            "mean_xyz": [round(float(v), 4) for v in vectors.mean(axis=0)],
+                            "std_xyz": [round(float(v), 4) for v in vectors.std(axis=0)],
+                            "mean_magnitude": round(float(magnitudes.mean()), 4),
+                            "max_magnitude": round(float(magnitudes.max()), 4),
```
[the remark this answers is not in the document]
Comment (vboussot) [reply konfai-mcp/konfai_mcp/file_io.py]: Fixed in 01cf74d: dense grids are summarised over bounded vector-aligned slices (_dense_displacement_summary) without materializing the dataset; a multi-chunk test pins the sliced stats to the single-pass values.

File: konfai-mcp/konfai_mcp/server.py [hunk 5]
```diff
@@ -1878,6 +1889,19 @@ def initialize_session(
...
+    """Refuse session management when a host (e.g. KonfAI Studio) pinned this server's session.
+
+    With KONFAI_MCP_SESSION set, the host owns the session<->UI binding: a session created or
+    switched here would run jobs in a workspace the host's panels never look at.
+    """
+    if os.environ.get("KONFAI_MCP_SESSION"):
+        raise ValueError(
+            f"This server is pinned to session '{WORKSPACE_LAYOUT.current_session}' by its host "
+            f"(KONFAI_MCP_SESSION): {action} is managed by the host UI. Keep working in the current session."
+        )
```
[the remark this answers is not in the document]
Comment (vboussot) [reply konfai-mcp/konfai_mcp/server.py:1902]: Not applied. ValueError is this package's refusal idiom (131 raise ValueError across konfai_mcp, surfaced by FastMCP as tool errors), and ConfigError is core KonfAI's YAML-configuration error: a pinned session is not a configuration problem. A lone ConfigError here would be the inconsistency.

File: konfai/data/patching.py [hunk 6]
```diff
@@ -2373,6 +2382,7 @@ def materialize_copies(
             )
         self._set_rewrite(rewrite)
         self.set_memory_budget(fallback_budget_bytes)
+        self._chain_device = device if device is not None and device.type != "cpu" else None
```
[the remark this answers is not in the document]
Comment (vboussot) [reply konfai/data/patching.py]: Fixed in 387ead4: the solo loop forwards device=device, and both materialize() and materialize_copies() scope _chain_device to the call with a context manager restored on every exit path. Tests cover the forwarding and the scoping.

File: README.md [hunk 7]
```diff
@@ -244,10 +244,10 @@ or a whole-volume statistic read once from disk (`Normalize`). On the stream
...
 with a peak resident set of 0.46 GiB.
 
-`konfai TRANSFORM` prints that decision per case *before* it writes a byte
-(STREAM, WHOLE-VOLUME, REDUCE or SKIP, naming the stage that refused to stream),
-and keeps the report at `./Transforms/<name>/plan.txt`. `--plan` prints the same
-report and stops without transforming.
+`konfai TRANSFORM` decides that per case *before* it writes a byte (STREAM,
+WHOLE-VOLUME, REDUCE or SKIP, naming the stage that refused to stream). The
+console gets a one-line summary; the run's log opens with the plan in full.
+`--plan` prints that full report and stops without transforming.
```
[the remark this answers is not in the document]
Comment (vboussot) [reply README.md]: Fixed in 7a3b2a1: the README lists STREAM/LOAD, WHOLE-VOLUME naming the refusing stage, REDUCE or REFUSED for a reduction, and SKIP as output-already-exists.

File: studio/frontend/src/App.tsx [hunk 8]
```diff
@@ -828,7 +845,9 @@ function Studio({ remote }: { remote: boolean }) {
             </button>
           </>
         ) : (
-          <span className="seg dim">local · offline · nothing leaves the machine</span>
+          <span className="seg dim">
+            {brain === "openai" ? "local · offline · nothing leaves the machine" : "local · your imaging data stays on this machine"}
+          </span>
```
[the remark this answers is not in the document]
Comment (vboussot) [reply studio/frontend/src/App.tsx:851]: Fixed in be330cd: the offline claim is gone; the status bar states what Studio can actually assert (imaging data stays on this machine).

File: studio/frontend/src/Chat.tsx [hunk 9]
```diff
@@ -590,6 +687,12 @@ export default function Chat({
                   onBlur={() => llm.modelText.trim() !== llm.model && llm.onModel(llm.modelText.trim())}
                 />
               )}
+              <LlmConnect
+                brain={llm.brain}
+                available={llm.brains.find((b) => b.id === llm.brain)?.available ?? false}
+                detail={llm.brains.find((b) => b.id === llm.brain)?.detail ?? ""}
+                onConnect={llm.onConnect}
+              />
```
[the remark this answers is not in the document]
Comment (vboussot) [reply studio/frontend/src/Chat.tsx:697]: Fixed in be330cd: unavailable backends stay selectable, so their Connect control is reachable to add credentials.

File: studio/frontend/src/FolderBrowser.tsx [hunk 10]
```diff
@@ -25,26 +25,47 @@ export default function FolderBrowser({
...
+        setTyped(d.path);
         setDirs(d.dirs);
         setFiles(d.files ?? []);
         setParent(d.parent);
         setError("");
       })
-      .catch((e) => setError(`Can't open this folder (${(e as Error).message}).`));
-  }, [path]);
+      .catch((e) => setError((e as Error).message));
+  }, [path, reload]);
```
[the remark this answers is not in the document]
Comment (vboussot) [reply studio/frontend/src/FolderBrowser.tsx:66]: Fixed in be330cd as suggested: listing state cleared before each request, stale responses aborted and ignored, AbortError not surfaced.

File: studio/frontend/src/RightPanel.tsx [hunk 11]
```diff
@@ -1260,6 +1346,7 @@ export default function RightPanel({
   const [stopping, setStopping] = useState(false);
   const [expand, setExpand] = useState<{ runKey: string; seriesKey: string } | null>(null); // curve-zoom modal
   const [validating, setValidating] = useState(false);
+  const [note, setNote] = useState(""); // a refused action says why, instead of looking like a no-op
```
[the remark this answers is not in the document]
Comment (vboussot) [reply studio/frontend/src/RightPanel.tsx:1354]: Fixed in be330cd: the session reset effect clears the note.

File: studio/konfai_studio/jobs.py [hunk 12]
```diff
@@ -96,6 +124,75 @@ def _pid_alive(pid: Any) -> bool:
...
+        created = _job_created(job)
+        if created >= newest.get(key, (float("-inf"), {}))[0]:
+            newest[key] = (created, job)
+    return {key: _claim_of(job) for key, (_, job) in newest.items()}
+
+
+def _status_for(job: dict[str, Any], log: Path | None) -> str:
+    """A job's status, read against the log it owns: the single reading every block of the feed makes."""
+    claim = _claim_of(job)
+    return claim.status_at(log.stat().st_mtime) if log is not None and log.is_file() else claim.status
```
[the remark this answers is not in the document]
Comment (vboussot) [reply studio/konfai_studio/jobs.py]: Fixed in f6fa151 at both sites: stat() wrapped in try/except OSError, _status_for falls back to claim.status, add() returns early when the log is gone.

File: studio/konfai_studio/paths.py [hunk 13]
```diff
@@ -51,17 +56,44 @@ def _session_dir(session: str) -> Path:
...
     target = _jail(_sessions_root(), name)
-    if target is not None and target.is_dir():
-        shutil.rmtree(target, ignore_errors=True)
+    if target is None:
+        return False
+    shutil.rmtree(target, ignore_errors=True)
+    studio_side = _jail(_workspace_root() / ".konfai_studio" / "sessions", name)
+    if studio_side is not None:
+        shutil.rmtree(studio_side, ignore_errors=True)
+    return not target.exists()
```
[the remark this answers is not in the document]
Comment (vboussot) [reply studio/konfai_studio/paths.py:99]: Fixed in f6fa151: _delete_workspace sanitizes the name before both _jail calls, and _Registry.load() normalizes every restored key (and the order list, deduplicated) through _sane_session.

File: studio/konfai_studio/registry.py [hunk 14]
```diff
@@ -228,20 +244,25 @@ async def remove(self, name: str) -> bool:
...
-                    await state.agent.__aexit__(None, None, None)
-                except Exception:
-                    pass
+            agent, state.agent = state.agent, None
+            if agent is not None:
+                with suppress(Exception):
+                    await agent.__aexit__(None, None, None)
+
+    async def close(self) -> None:
+        await self.close_agents()
```
[the remark this answers is not in the document]
Comment (vboussot) [reply studio/konfai_studio/registry.py:268]: Fixed in f6fa151 as suggested: snapshot iteration under self._create.

File: studio/konfai_studio/server.py [hunk 15]
```diff
@@ -454,6 +483,42 @@ async def remember_file(req: DatasetPath) -> dict[str, list[str]]:
...
+    try:
+        saved = json.loads(path.read_text(encoding="utf-8"))
+    except (OSError, json.JSONDecodeError):
+        saved = {}
+    saved.update(values)
+    path.write_text(json.dumps(saved), encoding="utf-8")
+    path.chmod(0o600)
+    for field, variable in _CREDENTIALS.items():
+        if field in values:
+            os.environ[variable] = values[field]
```
[the remark this answers is not in the document]
Comment (vboussot) [reply studio/konfai_studio/server.py:522]: Fixed in f6fa151: os.open(..., 0o600) at creation, parent directory created 0700, chmod kept for files written by older versions.

File: studio/tests/test_run_panels.py [hunk 16]
```diff
@@ -0,0 +1,307 @@
...
+def test_one_metric_has_one_direction_across_the_whole_session(workspace: Path) -> None:
+    """A direction is a property of the criterion, not of the run. Runs predating a criterion's
+    `maximize` flag declared the opposite, so the same PSNR read '↓' on one card and '↑' on the next,
+    each contradicting the leaderboard beside them."""
+    old = workspace / "AppEvaluations" / "eval_old" / "ImpactSynth"
+    new = workspace / "AppEvaluations" / "eval_new" / "ImpactSynth"
+    metric_file(old, directions={"PSNR": "min"}, means={"PSNR": 30.0})
+    metric_file(new, directions={"PSNR": "max"}, means={"PSNR": 35.0})
+    # The newest declaration wins, so touch the newer one last.
+    (new / "Metric_TRAIN.json").touch()
```
[the remark this answers is not in the document]
Comment (vboussot) [reply studio/tests/test_run_panels.py]: Fixed in f6fa151: explicit distinct mtimes via os.utime().

File: studio/tests/test_smoke.py [hunk 17]
```diff
@@ -56,6 +56,21 @@ def test_serves_the_built_frontend() -> None:
     assert "text/html" in response.headers.get("content-type", "")
 
 
+def test_a_workspace_created_outside_studio_joins_the_rail(tmp_path, monkeypatch: pytest.MonkeyPatch) -> None:
```
[the remark this answers is not in the document]
Comment (vboussot) [reply studio/tests/test_smoke.py]: Fixed in f6fa151.

File: tests/unit/test_case_expansion.py [hunk 18]
```diff
@@ -349,6 +349,32 @@ def test_a_region_draw_takes_its_own_pass_and_the_plan_names_the_draw(tmp_path:
     assert out.is_dataset_exist("CT", "CASE_000_r01") and out.is_dataset_exist("CT", "CASE_000_r02")
 
 
+def test_a_solo_copy_sweeps_on_the_requested_device(tmp_path: Path, monkeypatch) -> None:
```
[the remark this answers is not in the document]
Comment (vboussot) [reply tests/unit/test_case_expansion.py:352]: Done in #109 (3e01b18), which rewrites this file: `monkeypatch: pytest.MonkeyPatch`.
